### PR TITLE
Cleanup the s_time command

### PR DESCRIPTION
**Pull request by bernd-edlinger**
Various code-cleanups.
Handle WANT_READ/WRITE/CONNECT like other errors,
because all sockets are blocking, and this should not happen,
Turn off the linger option on connected sockets to avoid failure.
Continue test even without -cipher option as in 1.0.2.

### CODE REVIEW

File: apps/s_time.c [hunk 1]
```diff
@@ -240,16 +224,10 @@ int s_time_main(int argc, char **argv)
                                    www_path);
             if (buf_len <= 0 || SSL_write(scon, buf, buf_len) <= 0)
                 goto end;
-            while ((i = SSL_read(scon, buf, sizeof(buf))) > 0 ||
-                        SSL_get_error(scon, i) == SSL_ERROR_WANT_READ ||
-                        SSL_get_error(scon, i) == SSL_ERROR_WANT_WRITE)
```
Comment (mattcaswell) [inline apps/s_time.c]: This code only just got added recently. Why are you taking it out again?

Comment (bernd-edlinger) [inline apps/s_time.c]: yeah, sorry that I did not notice earlier, but the sockets here are all blocking.
implementing non-blocking sockets would be much more effort.

Comment (bernd-edlinger) [inline apps/s_time.c]: for instance the SSL_write above...

File: apps/s_time.c [hunk 2]
```diff
@@ -383,8 +347,10 @@ static SSL *doConnection(SSL *scon, const char *host, SSL_CTX *ctx)
 {
     BIO *conn;
     SSL *serverCon;
-    int width, i;
-    fd_set readfds;
+    int i;
+#if defined(SOL_SOCKET) && defined(SO_LINGER)
+    struct linger no_linger;
+#endif
```
Comment (dot-asm) [inline apps/s_time.c]: Why not make portion that makes the call to a block and declare no_linger there?

Comment (dot-asm) [inline apps/s_time.c]: I mean this way there would be one #ifdef...

File: apps/s_time.c [hunk 3]
```diff
@@ -433,6 +380,13 @@ static SSL *doConnection(SSL *scon, const char *host, SSL_CTX *ctx)
         return NULL;
     }
 
+#if defined(SOL_SOCKET) && defined(SO_LINGER)
+    no_linger.l_onoff  = 1;
+    no_linger.l_linger = 0;
+    (void) setsockopt(SSL_get_fd(serverCon), SOL_SOCKET, SO_LINGER,
+                      &no_linger, sizeof(no_linger));
```
Comment (dot-asm) [inline apps/s_time.c]: Formally speaking it might be more appropriate to implement it as BIO_CTRL. I mean whole point with BIO is to arrange implementation-neutral interface, and this kind of very specific. On the other hand this is special case...

File: apps/s_time.c (hunk 1 above)
Comment (mattcaswell) [inline apps/s_time.c]: I'm not sure I understand. SSL_ERROR_WANT_READ can occur with blocking sockets - and does so in the TLSv1.3 case (which is why this change was made).

Comment (bernd-edlinger) [inline apps/s_time.c]: Apparently I have not tried with TLSv1.3, sorry.
Can you please explain why the SSL not simply reads the data?

Comment (bernd-edlinger) [inline apps/s_time.c]: I mean this is obviously an API change, isn't it?

Comment (mattcaswell) [inline apps/s_time.c]: > Can you please explain why the SSL not simply reads the data?

Because there is no data to read. The arrival of the NewSessionTicket message means that a record gets processed by libssl, but it does not yield any application data.

> I mean this is obviously an API change, isn't it?

Nope. This has always been the case - its just been made more obvious by TLSv1.3. For example in TLSv1.2 this could happen if a reneg occurs.

Comment (mattcaswell) [inline apps/s_time.c]: E.g. in the SSL_read() man page:

> If the underlying BIO is blocking, a read function will only return once the read operation has been finished or an error occurred, except when a renegotiation takes place, in which case a SSL_ERROR_WANT_READ may occur.

That text probably needs some updating for TLSv1.3. Historically it could also occur if a heartbeat message is received (but that's been removed now). I think(?) it will also occur if a warning alert is received. Basically if any record arrives that doesn't yield application data.

File: apps/s_time.c (hunk 3 above)
Comment (bernd-edlinger) [inline apps/s_time.c]: Yes, kids, don't do that at home ;-)

File: apps/s_time.c (hunk 2 above)
Comment (bernd-edlinger) [inline apps/s_time.c]: done.

File: apps/s_time.c (hunk 1 above)
Comment (bernd-edlinger) [inline apps/s_time.c]: Hmm, I see.

My gut feeling this ERROR_WANT_READ did not happen without
the other side doing an explicit SSL_renegotiate, which is rarely the case,
and it is likely to break some more code out there.

I added SSL_CTX_set_mode(ctx, SSL_MODE_AUTO_RETRY)
but I got a new problem, the renegotiation does not work with TLS 1.3.

Could you have a look:
I build with ./config -d no-pic enable-tls1_3
./openssl s_server
sing default temp DH parameters
ACCEPT
ERROR
shutting down SSL
CONNECTION CLOSED
ACCEPT

./openssl s_time -reuse
No CIPHER specified


Now timing with session id reuse.
starting
ERROR

Comment (mattcaswell) [inline apps/s_time.c]: There are two problems here. The first is a bug in libssl (fix is in #3954). The second is slightly less obvious. Once the libssl bug is fixed this should make s_time with "-reuse" appear to work - but it won't actually be doing resumption at all! The reason is that it is not waiting for the NewSessionTicket to arrive before trying to use the session. See the "Sessions" section of my blog post for details: https://www.openssl.org/blog/blog/2017/05/04/tlsv1.3/

This is actually quite problematic. If we are talking to an OpenSSL server then we can be confident that the NewSessionTicket will arrive quite quickly. However other servers may not do so (or in theory they may not send one at all). For example, IIRC picotls doesn't send the NewSessionTicket until the server sends some application data - which might be some time (or not at all). At least that's the way it used to work - I've not tried testing against it recently.

Comment (bernd-edlinger) [inline apps/s_time.c]: Well, I re-based and squashed the s_time cleanup after your patch.
I expected at least the -www option to do real session renegotiation.
but that does probably not happen because the number of connects
per second seems to be no different with -reuse:

./openssl s_server -www

./openssl s_time -www /test -reuse
vs.
./openssl s_time -www /test -new

Comment (bernd-edlinger) [inline apps/s_time.c]: I mean sending "GET /blabla HTTP/1.1...."
and receiving "<BODY><HTML>" should be enough for session?

Comment (mattcaswell) [inline apps/s_time.c]: Yes. That should be enough. You might want to check what SSL_session_reused() says.

Comment (bernd-edlinger) [inline apps/s_time.c]: weird...

./openssl s_server -www
vs.
./openssl s_time -new
=> 
5134 connections in 8.30s; 618.55 connections/user sec, bytes read 0
5134 connections in 31 real seconds, 0 bytes read per connection
./openssl s_time -reuse
=> (as expected SSL_session_reused() = 0)
5049 connections in 8.36s; 603.95 connections/user sec, bytes read 0
5049 connections in 31 real seconds, 0 bytes read per connection

./openssl s_time -www /test -new
=>
4435 connections in 8.31s; 533.69 connections/user sec, bytes read 21953250
4435 connections in 31 real seconds, 4950 bytes read per connection

Note: the connections/user sec are almost identical, does that mean the client
has no less work to do for a reuse, but the server ?

./openssl s_time -www /test -reuse
=> (as expected SSL_session_reused() = 1)
8734 connections in 14.90s; 586.17 connections/user sec, bytes read 38840098
8734 connections in 31 real seconds, 4447 bytes read per connection

now I start the server with tls1_2
./openssl s_server -www -tls1_2

./openssl s_time -new
=>
4626 connections in 7.92s; 584.09 connections/user sec, bytes read 0
4626 connections in 31 real seconds, 0 bytes read per connection

./openssl s_time -reuse
=>
40253 connections in 16.34s; 2463.46 connections/user sec, bytes read 0
40253 connections in 31 real seconds, 0 bytes read per connection
Note: *This* is the only data point which is how I would expect it. 

./openssl s_time -www /test -new
=>
4527 connections in 7.85s; 576.69 connections/user sec, bytes read 22413177
4527 connections in 31 real seconds, 4951 bytes read per connection

./openssl s_time -www /test -reuse
=> (yes SSL_session_reused() = 1 here)
725 connections in 0.87s; 833.33 connections/user sec, bytes read 2657125
725 connections in 31 real seconds, 3665 bytes read per connection


Is this the Nagle algorithm at work???

Comment (bernd-edlinger) [inline apps/s_time.c]: Yes, adding NODELAY to the connect socket helps.
But TLS1.2 -renew still outperforms TLS1.3...

Comment (bernd-edlinger) [inline apps/s_time.c]: Hmm...

I tried this but it did not help:
```
diff --git a/apps/s_socket.c b/apps/s_socket.c
index 804ab5b..5b0e141 100644
--- a/apps/s_socket.c
+++ b/apps/s_socket.c
@@ -179,7 +179,8 @@ int do_server(int *accept_sock, const char *host, const char
     asock = BIO_socket(BIO_ADDRINFO_family(res), BIO_ADDRINFO_socktype(res),
                        BIO_ADDRINFO_protocol(res), 0);
     if (asock == INVALID_SOCKET
-        || !BIO_listen(asock, BIO_ADDRINFO_address(res), BIO_SOCK_REUSEADDR)) {
+        || !BIO_listen(asock, BIO_ADDRINFO_address(res),
+                       BIO_SOCK_REUSEADDR | BIO_SOCK_NODELAY)) {
         BIO_ADDRINFO_free(res);
         ERR_print_errors(bio_err);
         if (asock != INVALID_SOCKET)
```

still tls1.2 -www -reuse
28410 connections in 13.43s; 2115.41 connections/user sec, bytes read 104177885
28410 connections in 31 real seconds, 3666 bytes read per connection

and tls1.3 -www -reuse
8615 connections in 15.03s; 573.19 connections/user sec, bytes read 38285060
8615 connections in 31 real seconds, 4444 bytes read per connection

File: apps/s_time.c [hunk 4]
```diff
@@ -432,6 +378,16 @@ static SSL *doConnection(SSL *scon, const char *host, SSL_CTX *ctx)
             SSL_free(serverCon);
         return NULL;
     }
+#if defined(SOL_SOCKET) && defined(SO_LINGER)
+    else {
```
Comment (dot-asm) [inline apps/s_time.c]: You don't really need `else` here. Starting with straight `{` works and I'd say is more readable in this case.

Comment (bernd-edlinger) [inline apps/s_time.c]: well, a matter of taste, but if you like it better without else it's ok for me.

Comment (dot-asm) [inline apps/s_time.c]: Some even "hate the else after return" :-) But to me, in this case, it's really `else` in #if. I'm not actually a stranger to such construct (I mean I've used it myself), but it's always "unless actually necessary". And it's not actually necessary in this case...

File: apps/s_time.c (hunk 1 above)
Comment (bernd-edlinger) [inline apps/s_time.c]: so my conclusion is that reused sessions in tls1.3
cause the same cpu cycles for the client than new tls1.3 sessions,
just the server has significantly less cpu cycles.
OTOH tls1.2 session reuse may cause more round trips,
but much less cpu load on both server and client, and if
the network latency is very low as in this case the numbers
are not as convincing as they should be.
Is this by design, or could the TLS session have an
estimation of the network delay and choose another
strategy, if the round trips are cheap?

Comment (mattcaswell) [inline apps/s_time.c]: > ./openssl s_server -www
vs.
./openssl s_time -new
=>
5134 connections in 8.30s; 618.55 connections/user sec, bytes read 0
5134 connections in 31 real seconds, 0 bytes read per connection
./openssl s_time -reuse
=> (as expected SSL_session_reused() = 0)
5049 connections in 8.36s; 603.95 connections/user sec, bytes read 0
5049 connections in 31 real seconds, 0 bytes read per connection

These numbers are as expected, i.e. we get less connections/sec with the attempted reuse than without. In both cases we are doing a full handshake (we're not actually reusing)...but we've gone through some of the hoops of attempting to reuse. At least that's my guess.


> ./openssl s_time -www /test -new
=>
4435 connections in 8.31s; 533.69 connections/user sec, bytes read 21953250
4435 connections in 31 real seconds, 4950 bytes read per connection

Also expected. Less connections/sec because now we are actually transferring data.

> Note: the connections/user sec are almost identical, does that mean the client
has no less work to do for a reuse, but the server ?

I'm not sure what you mean. So far in the above numbers no reuse has actually happened (SSL_session_reused() == 0)


> ./openssl s_time -www /test -reuse
=> (as expected SSL_session_reused() = 1)
8734 connections in 14.90s; 586.17 connections/user sec, bytes read 38840098
8734 connections in 31 real seconds, 4447 bytes read per connection

For the first time connections/sec goes up because we are now really doing reuse, so this is as expected.

> now I start the server with tls1_2
./openssl s_server -www -tls1_2

> ./openssl s_time -new
=>
4626 connections in 7.92s; 584.09 connections/user sec, bytes read 0
4626 connections in 31 real seconds, 0 bytes read per connection

> ./openssl s_time -reuse
=>
40253 connections in 16.34s; 2463.46 connections/user sec, bytes read 0
40253 connections in 31 real seconds, 0 bytes read per connection
Note: This is the only data point which is how I would expect it.

> ./openssl s_time -www /test -new
=>
4527 connections in 7.85s; 576.69 connections/user sec, bytes read 22413177
4527 connections in 31 real seconds, 4951 bytes read per connection

All of the above are largely as expected. Session reuse is faster than no session reuse. TLSv1.3 without session reuse is out performing TLSv1.2 without session reuse.

Slightly surprising is that TLSv1.2 www without session reuse is marginally better than the TLSv1.3 equivalent.

> ./openssl s_time -www /test -reuse
=> (yes SSL_session_reused() = 1 here)
725 connections in 0.87s; 833.33 connections/user sec, bytes read 2657125
725 connections in 31 real seconds, 3665 bytes read per connection

This one I cannot explain. TLSv1.2 www with session reuse is much worse than without???

Comment (mattcaswell) [inline apps/s_time.c]: > Slightly surprising is that TLSv1.2 www without session reuse is marginally better than the TLSv1.3 equivalent.

Perhaps this could be to do with the cipher that is being selected in each case?

Comment (bernd-edlinger) [inline apps/s_time.c]: > 
> 
>     ./openssl s_time -www /test -reuse
>     => (yes SSL_session_reused() = 1 here)
>     725 connections in 0.87s; 833.33 connections/user sec, bytes read 2657125
>     725 connections in 31 real seconds, 3665 bytes read per connection
> 
> This one I cannot explain. TLSv1.2 www with session reuse is much worse than without???

Yes, that was before I added the NODELAY flags.
with NODELAY the same test results in much better numbers:

> still tls1.2 -www -reuse
> 28410 connections in 13.43s; 2115.41 connections/user sec, bytes read 104177885
> 28410 connections in 31 real seconds, 3666 bytes read per connection

that was the only data point that changed due to NODELAY.

**Comment by kaduk** [comment]
> Is this by design, or could the TLS session have an
estimation of the network delay and choose another
strategy, if the round trips are cheap?

I would prefer to not go down that route.  TLS 1.3 is pretty finalized, and the extra crypto has beneficial security properties.

**Comment by richsalz** [comment]
Generic TLS is not a time protocol. :) Does it even make sense to use TLS 1.3 for this, since some timestamps were removed?  Can we just error-out if only TLS 1.3 is available?

**Comment by bernd-edlinger** [comment]
@mattcaswell I wonder if this return WANT_READ thing for TLS1.3 will not
break a few non-blocking legacy applications, because previously that was
possible, but only when the server initiated a renegotiation which is rarely done
and if it is done it will not be right in the beginning of the conversation.

So I have two questions:
1. what is the benefit from returning WANT_READ for blocking sockets?
2. wouldn't it be better to implicitly enable SSL_MODE_AUTO_RETRY if
    TLS1.3 gets negotiated?

**Comment by mattcaswell** [comment]
> 1. what is the benefit from returning WANT_READ for blocking sockets?
> 2. wouldn't it be better to implicitly enable SSL_MODE_AUTO_RETRY if
TLS1.3 gets negotiated?

So, when writing with blocking sockets you may actually be seeking to write your code in such a way as to avoid blocking unnecessarily. This is actually how s_client/s_server are written. We only actually attempt to read data when we *know* there is data on the socket waiting to be read. If the socket indicates it is readable then we go and read it, otherwise we loop around waiting for some potential keyboard input. In some cases though (and particularly this happens in TLSv1.3 NST), the socket has data to be read but when you read it it doesn't actually yield any application data. In that case signalling WANT_READ is the right thing to do. Implicitly enabling SSL_MODE_AUTO_RETRY is likely to break such applications so I think it is better to stick with the documented behaviour.

**Comment by mattcaswell** [comment]
What is the status of this PR?

**Comment by bernd-edlinger** [comment]
I'll re-base.
I was not really satisfied with the performance numbers,
TLS1.2 was seemingly outperforming TLS 1.3 as far as I remember...

**Comment by bernd-edlinger** [comment]
Rebased.

I had to include `<internal/sockets.h>` because apparently socket headers were removed from e_os.h.

I think it was somewhat controversial that I prefer SSL_MODE_AUTO_RETRY
over handling SSL_ERROR_WANT_READ in a loop.
But I think this is fine, and we have not much test coverage for SSL_MODE_AUTO_RETRY
anyways.

**Comment by bernd-edlinger** [comment]
The change log mentions:
"Continue test even without -cipher option as in 1.0.2."
But that part was already fixed by another PR.
I would just update the commit message in a moment.

**Comment by bernd-edlinger** [comment]
Hmm.  removed 1.1.0 label for now, I think that won't cherry-pick.
at least the part with missing cipher option is not fixed in 1.1.0

**Comment by bernd-edlinger** [comment]
Merged to master, thanks!
